src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteExecutorKeepalived.java

```diff
@@ -0,0 +1,325 @@
+        // retry more than --remote_retry times which is not expected.
+        response = retrier
+            .execute(() -> Utils.refreshIfUnauthenticated(this::execute, callCredentialsProvider),
+                executeBackoff);
+
+        // If no response from Execute(), use WaitExecution() in a "loop" which is implemented
+        // inside the retry block.
+        //
+        // The cases to exit the loop:
+        //   1. Received the final response.
+        //   2. Received a un-retriable gRPC error.
+        //   3. Received NOT_FOUND error where we will retry Execute() (by returning null).
+        //   4. Received consecutive retriable gRPC errors (up to max retry times).
+        if (response == null) {
+          response = retrier.execute(
+              () -> Utils.refreshIfUnauthenticated(this::waitExecution, callCredentialsProvider),
+              waitExecutionBackoff);
+        }
+      }
+
+      return response;
+    }
+
+    @Nullable
+    ExecuteResponse execute() throws IOException {
+      Preconditions.checkState(lastOperation == null);
+
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get().execute(request);
+        return handleOperationStream(operationStream);
+      } catch (StatusRuntimeException e) {
+        // If lastOperation is not null, we know the execution request is accepted by the server. In
+        // this case, we will fallback to WaitExecution() loop when the stream is broken.
+        if (lastOperation != null) {
+          // At this point we know that the waitExecutionBackoff is just reset, so call
+          // nextDelayMillis() here to ensure the expected retry times(--remote_retry + 1) for
+          // waitExecutionBackoff: once you reset it, the next call to nextDelayMillis() will not
+          // increase the counter.
+          waitExecutionBackoff.nextDelayMillis();
+
+          // By returning null, we are going to fallback to WaitExecution() loop.
+          return null;
+        }
+        throw new IOException(e);
+      }
+    }
+
+    @Nullable
+    ExecuteResponse waitExecution() throws IOException {
+      Preconditions.checkState(lastOperation != null);
+
+      WaitExecutionRequest request = WaitExecutionRequest.newBuilder()
+          .setName(lastOperation.getName())
+          .build();
+      try {
+        Iterator<Operation> operationStream = executionBlockingStubSupplier.get()
+            .waitExecution(request);
+        return handleOperationStream(operationStream);
+      } catch (StatusRuntimeException e) {
+        // A NOT_FOUND error means Operation was lost on the server, retry Execute().
+        //
+        // However, we only retry Execute() if executeBackoff should retry. Also increase the retry
+        // counter at the same time (done by nextDelayMillis()).
+        if (e.getStatus().getCode() == Code.NOT_FOUND && executeBackoff.nextDelayMillis() >= 0) {
```

Comment:
(hiding a side effect in the nextDelayMillis() call is not great, but I think this is the best possible way to express what you want)

Reply:
Yes. I will probably refactor the Backoff and Retrier in another PR.